piker/brokers/core.py

```diff
@@ -81,54 +204,198 @@
+        cancelled = cancel_scope.cancelled_caught
+        if cancelled:
+            log.warn("Quote query timed out after 3 seconds, retrying...")
+            # handle network outages by idling until response is received
+            quotes = await wait_for_network(partial(get_quotes, tickers))
+
+        postquote_start = time.time()
+        q_payloads = {}
+        for symbol, quote in quotes.items():
+            # FIXME: None is returned if a symbol can't be found.
+            # Consider filtering out such symbols before starting poll loop
+            if quote is None:
+                continue
+
+            if diff_cached:
+                # if cache is enabled then only deliver "new" changes
+                last = _cache.setdefault(symbol, {})
+                new = set(quote.items()) - set(last.items())
+                if new:
+                    log.info(
+                        f"New quote {quote['symbol']}:\n{new}")
+                    _cache[symbol] = quote
+                    for queue in tickers2qs[symbol]:
+                        q_payloads.setdefault(queue, {})[symbol] = quote
+            else:
+                for queue in tickers2qs[symbol]:
+                    q_payloads.setdefault(queue, {})[symbol] = quote
+
+        # deliver to each subscriber
+        if q_payloads:
+            for queue, payload in q_payloads.items():
+                try:
+                    await queue.put(payload)
+                except (
+                    # That's right, anything you can think of...
+                    trio.ClosedStreamError, ConnectionResetError,
+                    ConnectionRefusedError,
+                ):
+                    log.warn(f"{queue.peer} went down?")
+                    for qset in tickers2qs.values():
+                        qset.discard(queue)
+
+        req_time = round(postquote_start - prequote_start, 3)
+        proc_time = round(time.time() - postquote_start, 3)
+        tot = req_time + proc_time
+        log.debug(f"Request + processing took {tot}")
+        delay = sleeptime - tot
+        if delay <= 0:
+            log.warn(
+                f"Took {req_time} (request) + {proc_time} (processing) "
+                f"= {tot} secs (> {sleeptime}) for processing quotes?")
+        else:
+            log.debug(f"Sleeping for {delay}")
+            await trio.sleep(delay)
+
+
+async def start_quoter(
+    broker2tickersubs: dict,
+    clients: dict,
+    dtasks: set,  # daemon task registry
+    nursery: "Nusery",
+    stream: trio.SocketStream,
+) -> None:
+    """Handle per-broker quote stream subscriptions.
+
+    Spawns new quoter tasks for each broker backend on-demand.
+    Since most brokers seems to support batch quote requests we
+    limit to one task per process for now.
+    """
+    queue = StreamQueue(stream)  # wrap in a shabby queue-like api
+    log.info(f"Accepted new connection from {queue.peer}")
+    async with queue.stream:
```

Comment:
Hmm does this drop the connection with the client if all subscriptions are removed?
Is that desired?